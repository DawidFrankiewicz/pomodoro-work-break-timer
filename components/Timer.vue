<script setup>
const props = defineProps({
    config: {
        type: Object,
        required: true,
    },
});

const workTime = ref(props.config.workTime);
const breakTime = ref(props.config.breakTime);

const timer = ref(workTime.value);
const totalTimer = ref(0);
const isTimerRunning = ref(false);
const isWorkTime = ref(true);

let timerInterval;

watch(
    () => props.config,
    () => {
        resetTimer();
    }
);

onUnmounted(() => {
    stopTimer();
});

const displayTime = (timeInSeconds, displayHours = false) => {
    const hours = Math.floor(timeInSeconds / 3600);
    const minutes = Math.floor(timeInSeconds / 60);
    const seconds = timeInSeconds % 60;

    if (displayHours) {
        return (
            String(hours).padStart(2, '0') +
            ':' +
            String(minutes % 60).padStart(2, '0') +
            ':' +
            String(seconds).padStart(2, '0')
        );
    }

    return (
        String(minutes).padStart(2, '0') +
        ':' +
        String(seconds).padStart(2, '0')
    );
};

const resetTimer = () => {
    stopTimer();
    totalTimer.value = 0;
    timer.value = workTime.value;
    isWorkTime.value = true;
};

const startTimer = () => {
    isTimerRunning.value = true;
    timerInterval = setInterval(() => {
        if (timer.value <= 0) {
            if (isWorkTime.value) {
                startBreak();
            } else {
                startWork();
            }
        } else {
            timer.value--;
            totalTimer.value++;
        }
    }, 1000);
};

const stopTimer = () => {
    isTimerRunning.value = false;
    clearInterval(timerInterval);
};

const startBreak = () => {
    timer.value = breakTime.value;
    isWorkTime.value = false;
};

const startWork = () => {
    timer.value = workTime.value;
    isWorkTime.value = true;
};
</script>

<template>
    <div
        class="mx-auto flex min-h-screen max-w-[1920px] flex-col items-center justify-center"
    >
        <h1 class="py-6 text-6xl font-bold">Work Brake Timer</h1>
        <div class="flex flex-col items-center rounded-sm px-6 py-5">
            <span class="text-3xl font-bold">
                {{ displayTime(totalTimer, true) }}
            </span>
            <span class="text-9xl font-bold">
                {{ displayTime(timer) }}
            </span>
        </div>
        <div class="rounded-sm px-6 py-5 text-3xl font-bold"></div>
        <div
            v-if="props.config.totalRounds"
            class="rounded-sm px-6 py-5 text-3xl font-bold"
        >
            Rounds: {{ props.config.totalRounds }}
        </div>
        <div
            v-if="props.config.isInfinite !== undefined"
            class="rounded-sm px-6 py-5 text-3xl font-bold"
        >
            Rounds: {{ props.config.isInfinite }}
        </div>

        <TimerControls
            :isTimerRunning="isTimerRunning"
            @startTimer="startTimer()"
            @stopTimer="stopTimer()"
        />
    </div>
</template>
6
