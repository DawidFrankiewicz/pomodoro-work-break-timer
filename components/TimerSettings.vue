<script setup>
const props = defineProps({
    defaultConfig: {
        type: Object,
        required: true,
    },
});

const emit = defineEmits(['update-config']);
const isConfigOpen = ref(false);
const config = ref(props.defaultConfig);
</script>

<template>
    <!-- Container -->
    <div
        class="fixed left-0 top-0 h-screen pl-6 pt-6 text-slate-900 transition-all hover:opacity-100"
        :class="isConfigOpen ? 'opacity-100' : '-translate-x-full opacity-40'"
    >
        <!-- Open button -->
        <button
            @click="isConfigOpen = !isConfigOpen"
            class="absolute right-0 top-12 h-16 w-16 translate-x-full rounded-e-md bg-purple-200 bg-gradient-to-bl from-purple-200 to-purple-400"
        >
            <font-awesome-icon :icon="['fas', 'gear']" class="fa-2xl" />
        </button>
        <!-- End Open button -->
        <!-- Config bar -->
        <div
            class="rounded-md bg-purple-200 bg-gradient-to-br from-purple-200 to-purple-400"
        >
            <div class="flex flex-col gap-2 p-4 text-slate-900">
                <h1 class="text-center text-xl font-bold">Timer Settings</h1>
                <div class="mb-2 flex flex-col gap-2">
                    <FormNumberInput
                        label="Work Time (Min)"
                        :value="config.workTime / 60"
                        :min="1"
                        @update="
                            (newValue) => (config.workTime = newValue * 60)
                        "
                    />
                    <FormNumberInput
                        label="Break Time (Min)"
                        :value="config.breakTime / 60"
                        :min="1"
                        @update="
                            (newValue) => (config.breakTime = newValue * 60)
                        "
                    />

                    <div>
                        <label class="block font-bold">Infinite</label>
                        <input type="checkbox" v-model="config.isInfinite" />
                    </div>
                    <div
                        class="transition-all duration-75"
                        :class="
                            config.isInfinite
                                ? 'pointer-events-none rounded-md bg-purple-200 p-2 opacity-40'
                                : ''
                        "
                    >
                        <FormNumberInput
                            label="Rounds"
                            :value="config.totalRounds"
                            :disabled="config.isInfinite"
                            :min="1"
                            @update="
                                (newValue) => (config.totalRounds = newValue)
                            "
                        />
                    </div>
                </div>
                <div class="flex min-h-[50px]">
                    <button
                        class="flex-grow rounded-t-md border-b-4 border-r-4 border-purple-800 bg-red-200 bg-gradient-to-br from-red-200 to-pink-300 px-4 text-xl font-bold shadow-sm transition-all duration-75 hover:ml-[2px] hover:mt-[2px] hover:border-b-2 hover:border-r-2"
                        @click="
                            $emit('update-config', Object.assign({}, config))
                        "
                    >
                        Apply
                    </button>
                </div>
            </div>
        </div>
        <!-- End Config bar -->
    </div>
    <!-- End Container -->
</template>
