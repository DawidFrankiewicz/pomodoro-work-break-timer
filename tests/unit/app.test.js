import { shallowMount, mount } from '@vue/test-utils';
import Timer from '@/components/Timer.vue';
import TimerSettings from '@/components/TimerSettings.vue';

const components = {
    Timer,
    TimerSettings,
};

describe('App.test.js', () => {
    test('Should render correctly', async () => {
        const component = await import('@/app.vue');
        expect(component).toBeDefined();

        const wrapper = shallowMount(component.default, {
            components: components,
        });

        // Check if component matches snapshot (HTML structure)
        // https://vitest.dev/guide/snapshot#updating-snapshots
        expect(wrapper.html()).toMatchSnapshot();
    });
});
