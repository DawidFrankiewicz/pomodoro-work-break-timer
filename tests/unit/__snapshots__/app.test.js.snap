// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App.test.js > Should render correctly 1`] = `
"<div data-v-938b83b0="" class="relative min-h-screen flex-grow bg-gray-100 bg-gradient-to-br from-gray-100 to-gray-200">
  <timer-stub data-v-938b83b0="" config="[object Object]"></timer-stub>
  <timer-settings-stub data-v-938b83b0="" defaultconfig="[object Object]"></timer-settings-stub>
</div>"
`;
